### ISSUE TITLE
Swap bitwise with logical operators

### ISSUE THREAD

**Issue by SheepTester** [chop0/aardvark#4]
The shorter variants should be used for logical operators rather than bitwise operators as the former is used more often, so

- `&` `|` `~` are the logical AND, OR, and NOT operators, respectively.
- `&&` `||` `!` are the bitwise AND, OR, and NOT operators, respectively.

`^` should also be reserved for exponentiation, so XOR can be something obscure like `^^`.

**Comment by ky28059** [comment]
Add logical xor and xand too

**Comment by chop0** [comment]
This is a good idea; we could maybe take it one step further and combine some of them.
  *`&` is logical AND in the case of comparing booleans, bitwise in terms of comparing bytes
  *`|` is logical OR in the case of booleans, bitwise for bytes
  *`~` or `!` is logical NOT for booleans, bitwise NOT otherwise

**Comment by SheepTester** [comment]
Ah, I see, kind of like how `+` is used for both concatenation and addition

**Comment by chop0** [comment]
@SheepTester I think types should be able to implement infix operators, like in rust.

**Comment by chop0** [comment]
test

**Comment by SheepTester** [comment]
> @SheepTester I think types should be able to implement infix operators, like in rust.

Rust docs say that [because and and or short-circuit, they don't have traits for them](https://doc.rust-lang.org/std/ops/index.html). However, at least for N, I think I'll make short circuiting a trait-defined thing. This has the benefit of not having to have something like `or` vs `or_else` for an optional/result monad.

**Comment by chop0** [comment]
@SheepTester Yeah, I'm going to include that in the next implementation of Aardvark; graal has made me mald too much, so I'm probably switching to rust or just vanilla java, at least for now.

**Comment by chop0** [comment]
On the other hand, I am a bit conflicted about making `||` kind of a special form; I'm going to be adding the concept of purity and pure functions to Aardvark, so I might make it so that `||` must have pure expressions on the left and right.

**Comment by SheepTester** [comment]
> On the other hand, I am a bit conflicted about making `||` kind of a special form; I'm going to be adding the concept of purity and pure functions to Aardvark, so I might make it so that `||` must have pure expressions on the left and right.

Hmm, I feel like short circuiting would be rather necessary, especially in cases where it's used in recursion (though I can't think of why it'd be used directly in recursion), unless you're going to make expressions evaluate lazily? :eyes:

**Comment by chop0** [comment]
I might evaluate pure functions lazily. What do you mean by recursion?

**Comment by SheepTester** [comment]
> I might evaluate pure functions lazily. What do you mean by recursion?

🙂 
```scheme
(define (any? pred ls)
  (or (pred (car ls)) (any? pred (cdr ls))))
```

**Comment by chop0** [comment]
![image](https://user-images.githubusercontent.com/8400440/103240667-d3ae1f00-4905-11eb-895a-7e2d50301fab.png)
🤔

**Comment by SheepTester** [comment]
Ah, explains 2020
